add `zfp` codec

LDeakin opened the pull request:
I intend to release this codec with the next [`zarrs`](https://crates.io/crates/zarrs) release. If this codec configuration and implementation is not controversial, it would be good if it were just named `zfp`?

Previous `zarrs` releases had some differences from this spec:
- `"name": "https://codec.zarrs.dev/array_to_bytes/zfp"`
- Supports a `write_header` parameter that I would prefer to remove

The inclusion of a `write_header` parameter might be worthy of discussion though. If it were present, then this codec is fully compatible with data encoded with `zfpy` in Zarr V2 / `numcodecs.zfpy` in Zarr V3 with just a change of metadata. I support `zfpy` independently in `zarrs`, although it is the same underlying codec implementation.

normanrz commented:
Thanks! This PR looks very good.
I don't see why this codec couldn't be registered as `zfp`.

jbms commented on codecs/zarrs.zfp/README.md:
```diff
@@ -0,0 +1,88 @@
+- `int32`, `uint32`, `int64`, `uint64`, `float32`, `float64`
+- `int8`, `uint8`, `int16`, `uint16` (through internal promotion 32-bit)
+
+Implementations are permitted to support additional data types that could be interpreted as the above data types (e.g. `datetime64` -> `uint64`).
```
I think datetime64 would correspond to int64 not uint64.

jbms replied:
You could note that lower-precision floating point types, such as float16 and the new float-precision floating point extension types I added, can be supported by first converting to float32 in the normal way.

Lower-precision integer types, e.g. {int,uint}{2,4} can also be supported but would need to be promoted in the same way that {int,uint}{8,16} are promoted:

for intN:

```c++
int32_value = (int32_t)intN_value << (31 - N)
intN_value = (intN_t)clamp(int32_value >> (31 - N), 1<<(N-1), (1<<(N-1)) - 1)
```

for uintN

```c++
int32_value = ((int32_t)uintN_value - (1<<(N-1))) << (31 - N)
uintN_value = (uintN_t)clamp((int32_value >> (31 - N)) + (1<<(N-1)), 0, (1<<N) - 1)
```

jbms commented on codecs/zarrs.zfp/README.md:
```diff
@@ -0,0 +1,88 @@
+# ZFP codec
+
+Defines a `array -> bytes` codec that compresses chunks using the [zfp](https://github.com/LLNL/zfp) algorithm.
+
```
zfp natively only supports 1, 2, 3 and 4 dimensional arrays.  However, for zarr it might be useful to allow arrays with more than 4 dimensions provided that at most 4 dimensions have a chunk size greater than 1.  That would allow zfp to be combined with sharding_indexed to handle arbitrary numbers of dimensions.  You would just need to define exactly how the zfp dimensions are determined from the zarr dimensions. (See also https://github.com/seung-lab/zfpc which is supported by the neuroglancer-precomputed format --- with zarr I think an extra container format is unnecessary because sharding_indexed already serves that purpose quite well)

jbms replied:
My suggestion for defining that would be to say: if the number of dimensions with chunk size > 1 is within the range `[1, 4]`, those are the zfp dimensions.    If there are more than 4 such dimensions, it is an error.  If there are zero such dimensions, i.e. the chunk contains just a single element, treat it as a 1-d array of 1 element (so that users don't have to worry about this as a special case).

LDeakin replied:
Great catch, and a good idea in regards to supporting higher dimensional arrays. I'll add support for that in `zarrs`.

I wasn't aware of https://github.com/seung-lab/zfpc, but it highlights the reason why I didn't include complex numbers for this codec. A Zarr V3 way to do that could be an array-to-bytes codec that deinterleaves complex data types and encodes them independently.

jbms replied:
Regarding your "may support" change:  what I described (and what you have) is a breaking change, because previously if you had a chunk size of [64, 64, 1, 64] it would get encoded as a 4-d zfp (effectively padded to [64, 64, 4, 64]) --- I'm not sure exactly what the implications are in terms of runtime performance and compression ratio, but it seems plausible that it may make compression and decompression 4x as slow.  Under the new language, it would instead get encoded as a 3-d zfp with chunk shape `[64, 64, 64]`.  Without the header I don't know that this can be detected automatically.  For zarrs, assuming backwards compatibility is important, perhaps you can support the old behavior under the old identifier that you used, and the new behavior under the plain `zfp` identifier.

LDeakin replied:
In that case, always maintaining the chunk shape for <=4D data and only collapsing dimensions for 5D+ data might be a good option. I'll trial this in implementation, and come back to this PR afterwards.

jbms commented:
At least based on my understanding, having a chunk size of 1 for one of the zfp dimensions is never actually desirable and you are retaining that solely for backwards compatibility.  Having different behavior for >4 dimensions where chunk size=1 dimensions get dropped may be rather confusing.

An alternative approach would be to delegate the dimension adjustments to a separate "array -> array" codec, perhaps called "squeeze" in line with `np.squeeze`, that allows dimensions of chunk size 1 to be dropped.   Dimension manipulations of various sorts will surely be needed to support various image formats as well, and doing that with common "array -> array" codecs rather than ad-hoc options as part of the "array -> bytes" codec may be better.

LDeakin commented:
> At least based on my understanding, having a chunk size of 1 for one of the zfp dimensions is never actually desirable

For sure, having just read through the algorithm.

> An alternative approach would be to delegate the dimension adjustments to a separate "array -> array" codec, perhaps called "squeeze" in line with `np.squeeze`, that allows dimensions of chunk size 1 to be dropped

I like that! If you are onboard with it, I will remove mention of collapsing dimensions in this codec, but keep a note about how higher dimensionality arrays could be supported with a `squeeze`-like codec.

LDeakin commented:
I added a `squeeze` codec to `zarrs` and updated this spec accordingly. I like this approach best for supporting higher-dimensional arrays. Thanks for all your comments @jbms.

normanrz commented:
Thanks for working on this! I'll merge this tomorrow unless I hear from you that you still want to iterate pre-merge.

normanrz commented:
Right after merging, I noticed that the `schema.json` and folder name still referenced `zarrs.zfp`. I changed that in https://github.com/zarr-developers/zarr-extensions/commit/77c6d6ce498754d3acdfdaf9901bd88366421861. I hope that is what you intended.
